pallet/staking/src/lib.rs

```diff
@@ -1208,3 +1235,126 @@ where
+/// ```
+pub struct KtonRewardDistributionOwner;
+impl<T> Get<T> for KtonRewardDistributionOwner
+where
+	T: From<[u8; 20]>,
+{
+	fn get() -> T {
+		[115, 99, 47, 107, 116, 115, 116, 107, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0].into()
+	}
+}
+
+/// A curve helps to migrate to staking v2 smoothly.
+pub struct MigrationCurve<T>(PhantomData<T>);
+impl<T> Get<Perquintill> for MigrationCurve<T>
+where
+	T: Config,
+{
+	fn get() -> Perquintill {
+		// substrate
+		use sp_runtime::traits::SaturatedConversion;
+
+		let x = (<frame_system::Pallet<T>>::block_number() - <MigrationStartBlock<T>>::get())
+			.saturated_into::<u64>()
+			.max(1);
+		let month_in_blocks = 30 * 24 * 60 * 60 / 12;
+
+		Perquintill::one() - Perquintill::from_rational(x, month_in_blocks)
+	}
+}
+
+/// KTON staker contact notification interface.
+pub trait KtonStakerNotification {
+	/// Notify the KTON staker contract.
+	fn notify(_: Balance) {}
+}
+impl KtonStakerNotification for () {}
+/// KTON staker contact notifier.
+pub struct KtonStakerNotifier<T>(PhantomData<T>);
+impl<T> KtonStakerNotification for KtonStakerNotifier<T>
+where
+	T: Config + darwinia_message_transact::Config,
+	T::RuntimeOrigin: Into<Result<LcmpEthOrigin, T::RuntimeOrigin>> + From<LcmpEthOrigin>,
+	<T as frame_system::Config>::AccountId: Into<H160>,
+{
+	fn notify(amount: Balance) {
+		let Some(signature)= mock_sig() else {
+			log::error!("[pallet::staking] Invalid mock signature for the staking notify transaction.");
+
+			return;
+		};
+		// KTONStakingRewards
+		// 0x000000000419683a1a03AbC21FC9da25fd2B4dD7
+		let staking_reward =
+			H160([0, 0, 0, 0, 4, 25, 104, 58, 26, 3, 171, 194, 31, 201, 218, 37, 253, 43, 77, 215]);
+		// RewardsDistribution Contract
+		// 0x000000000Ae5DB7BDAf8D071e680452e33d91Dd5
+		let reward_distr = H160([
+			0, 0, 0, 0, 10, 229, 219, 123, 218, 248, 208, 113, 230, 128, 69, 46, 51, 217, 29, 213,
+		]);
+		// https://github.com/darwinia-network/kton-staker/blob/175f0ec131d4aef3bf64cfb2fce1d262e7ce9140/src/RewardsDistribution.sol#L11
+		#[allow(deprecated)]
+		let function = Function {
+			name: "distributeRewards".into(),
+			inputs: vec![
+				Param {
+					name: "ktonStakingRewards".into(),
+					kind: ParamType::Address,
+					internal_type: None,
+				},
+				Param { name: "reward".into(), kind: ParamType::Uint(256), internal_type: None },
+			],
+			outputs: vec![Param {
+				name: "success or not".into(),
+				kind: ParamType::Bool,
+				internal_type: None,
+			}],
+			constant: None,
+			state_mutability: StateMutability::Payable,
+		};
+
+		let notify_transaction = LegacyTransaction {
+			nonce: U256::zero(),     // Will be reset in the message transact call
+			gas_price: U256::zero(), // Will be reset in the message transact call
+			gas_limit: U256::from(1_000_000), /* It should be big enough for the evm
+			                          * transaction, otherwise it will out of gas. */
+			action: TransactionAction::Call(reward_distr),
+			value: U256::zero(),
```

Comment:
Since we pass zero value here, the value should be sent directly to reward_distr when reward.